[blog] Add Material UI v6 stable release (#41932)
mui/material-ui: docs/pages/blog/material-ui-v6-sneak-peek.md

```jsx
const Component = styled('div')(({ theme }) => ({
  display: 'flex',
  flexDirection: 'column',
  gap: theme.spacing(2),
  [theme.containerQueries.up('sm')]: {
    flexDirection: 'row',
  },
  [theme.containerQueries('sidebar').up('400px')]: {
    flexDirection: 'row',
  },
}));
```

It also works with the `sx` prop:

```jsx
<Card
  sx={{
    '@sm': {
      flexDirection: 'row',
    },
    '@400/sidebar': {
      flexDirection: 'row',
    },
  }}
/>
```

## Stablizing CSS Theme Variables

We introduced [CSS theme variables](https://mui.com/material-ui/experimental-api/css-theme-variables/overview/) in v5 as an experimental API. In v6, we are making it a stable API with two more additions:

### Spacing

The value from `theme.spacing()` will use the CSS variable `--mui-spacing` under the hood including `sx` values.

```jsx
<Button
  sx={(theme) => ({
    padding: theme.spacing(2), // calc(var(--mui-spacing) * 2)
  })}
/>
```

### Typography

The typography tokens will be generated as CSS variables for using with CSS `font` property.

You can access the typography tokens from `theme.vars.typography.*`:

```jsx
const Title = styled('div')(({ theme }) => ({

>   font: theme.vars.typography.h1,

Reviewer: ```suggestion
  fontFamily: theme.vars.typography.h1.fontFamily,
```
Or

```suggestion
  ...theme.vars.typography.h1,
```

}));
```

or refer to the CSS variable directly:

```css
.some-component {

>   font: var(--mui-typography-h1);

Reviewer: This'll also change

---

> title: Material UI v6 Sneak Peek

Reviewer 2: Maybe this change better communicates that v6 is close in time.

```suggestion
title: Material UI v6 now in beta
```

Author: Changed to stable release. I will let @samuelsycamore casts his magic for the title.

description: Previewing all the major updates of Material UI v6.
date: 2024-07-01T00:00:00.000Z

> authors: ['siriwatknp']

Reviewer 2: Not sure which is the standard, but per [this comment](https://github.com/mui/material-ui/pull/42198/files#r1602631793) I guess we are missing some people here.

Author: Yes, I will add everyone involved in v6.

tags: ['Material UI', 'Product']
manualCard: true
---

> Material UI v6 is now in beta! We're excited to share a sneak peek of all the upcoming major updates.

Reviewer 2: Not sure if "sneak peek" is the right term because users can already try these features.

```suggestion
Material UI v6 is now in beta! We're excited to share all the upcoming major updates.
```

## Opt-in CSS extraction via Pigment CSS

> As you know, Material UI v5 is using Emotion as a default styling solution. As a runtime CSS-in-JS library, it has several trade-offs such as slower performance and larger bundle size.

Reviewer 2: ```suggestion
As you know, Material UI v5 uses Emotion as a default styling solution. As a runtime CSS-in-JS library, it has several trade-offs such as slower performance and larger bundle size.
```

> With Material UI v6, we're introducing an opt-in integration with Pigment CSS, our new open-source library, that will eliminate the overhead of style recalculation while preserving similar APIs that you are already familiar with.

Reviewer 2: Link to Pigment CSS repo?

```suggestion
With Material UI v6, we're introducing an opt-in integration with [Pigment CSS](https://github.com/mui/pigment-css), our new open-source library, that will eliminate the overhead of style recalculation while preserving similar APIs that you are already familiar with.
```

> Since the integration is an opt-in feature, you can upgrade to v6 and continue using Emotion if you'd like. We will provide a migration guide to help you switch to Pigment CSS at your pace.

Reviewer 2: ```suggestion
Since the integration is an opt-in feature, you can upgrade to v6 and continue using Emotion if you'd like. We will provide a migration guide to help you switch to Pigment CSS at your own pace.
```

<!-- would be nice if we have a link to a prototype page -->
<!-- a link to Next.js repo that render templates -->

### React Server Components

Once integrated with Pigment CSS, Material UI v6 provides a separate set of layout components, including `Grid`, `Container`, and `Stack`, that are compatible with React Server Components.

```jsx
import Grid from '@mui/material/PigmentGrid';
import Container from '@mui/material/PigmentContainer';
import Stack from '@mui/material/PigmentStack';
```

<!-- A link to example repo? -->

## CSS theme variables

> A new API that lets you turn your theme into CSS variables.

Reviewer 2: Should we rephrase this to say that the experimental `CssVarsProvider` has been stabilized instead of saying "a new api"?

```diff
- import { ThemeProvider } from '@mui/material/styles';
+ import { CssVarsProvider } from '@mui/material/styles';
```

Serializable values like palette, spacing, typography, etc. are converted to CSS variables and can be accessed from `theme.vars`.

> With CSS variables, you can easily integrate your prefered styling solution with Material UI.

Reviewer 2: ```suggestion
With CSS variables, you can easily integrate your preferred styling solution with Material UI.
```

}
```

```css Result
:root {
  --palette-background-default: #fff;
  /* ...other variables */
}

:where([data-mode-dark]) {
  --palette-background-default: #121212;
  /* ...other variables */
}
```

</codeblock>

### Custom styles for specific modes

The new API `theme.applyStyles()` has been added to unify all supporting styling solutions, including Emotion, Styled-components and Pigment CSS.

We will provide a codemod that migrates the code like an example below:

```diff
  const StyledInput = styled(InputBase)(({ theme }) => ({
    padding: 10,
    width: '100%',
-    borderBottom: `1px solid ${
-      theme.palette.mode === 'dark' ? '#30363d' : '#eaecef'
-    }`,
+    borderBottom: `1px solid #eaecef`,
+    ...theme.applyStyles('dark', {
+      borderBottom: '1px solid #30363d',
+    })
    '& input': {
      borderRadius: 4,
-      backgroundColor: '#fff',
+     ...theme.applyStyles('dark', {
+       backgroundColor: '#0d1117',
+     })
    },
  }));
```

## Container Queries

We added support for [container queries](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_containment/Container_queries) based on the existing `theme.breakpoints` API.

> This feature lets you define styles based on the width of the parent container instead of the viewport.

Reviewer 2: ```suggestion
This feature lets you define styles based on the parent container's width instead of the viewport.
```

```jsx
const Component = styled('div')(({ theme }) => ({
  display: 'flex',
  flexDirection: 'column',
  gap: theme.spacing(2),
  [theme.containerQueries.up('sm')]: {
    flexDirection: 'row',
  },
  [theme.containerQueries('sidebar').up('400px')]: {
    flexDirection: 'row',
  },
}));
```

It also works with the `sx` prop:

```jsx
<Card
  sx={{
    '@sm': {
      flexDirection: 'row',
    },
    '@400/sidebar': {
      flexDirection: 'row',
    },
  }}
/>
```

## Stabilized Grid v2

> ## Revamping the free templates

Reviewer 3: ```suggestion
## Revamping the free templates

Explore the new and enhanced [Material UI free templates](https://mui.com/material-ui/getting-started/templates/) to see these amazing features in action. We've fully revamped the templates to provide the perfect starting point for your project, whether you're adding sleek styles or using the template's sections.

<!-- First demo  -->

### Custom syles

The new custom styles we've added allow you to dive into the best approaches for customization or simply copy and paste some stylish elements.

<!-- Second demo -->

Stay tuned, as many more updates are coming in the upcoming months! We'd love to hear your feedback on which templates or sections you'd like to see next.
```

### First demo

For the first demo, I thought on having something like a carousel, showcasing the new templates, with links for the preview and source code.

![image](https://github.com/mui/material-ui/assets/37222944/e2812982-763d-4779-be70-64b3cd553ee6)

👉 [Figma prototype](https://www.figma.com/design/4uv2dT18rXJPZBbrbpw61q/Marketing-site?node-id=10026-123&t=Jum9nq4wYDtxx1Cd-11)

### Second demo

For the second, I thought on having something like we had for color inversion on Joy UI marketing page:

![image](https://github.com/mui/material-ui/assets/37222944/fb8d58b2-6994-491b-8846-009d7cd30faf)

👉 [Figma prototype](https://www.figma.com/design/4uv2dT18rXJPZBbrbpw61q/Marketing-site?node-id=10026-124&t=Jum9nq4wYDtxx1Cd-11)

Reference:

<img width="400" alt="Screenshot 2024-07-09 at 09 50 03" src="https://github.com/mui/material-ui/assets/37222944/fa31b62b-2e84-447d-9a0b-0b3dd520441e">

Author: Got it!
